diff --git a/bfd/elf32-m68k.c b/bfd/elf32-m68k.c
index 16aa5918518..5f70f174e1c 100644
--- a/bfd/elf32-m68k.c
+++ b/bfd/elf32-m68k.c
@@ -3789,12 +3789,29 @@ elf_m68k_relocate_section (bfd *output_bfd,
 
 	  break;
 
-	case R_68K_8:
-	case R_68K_16:
-	case R_68K_32:
 	case R_68K_PC8:
 	case R_68K_PC16:
 	case R_68K_PC32:
+	  if (sec != NULL && sec->output_section != NULL
+	      && (sec->output_section->flags & SEC_READONLY) == 0)
+	    {
+	      unsigned char code;
+	      bfd_byte *ptr;
+
+	      /* TODO: Verify the op is valid for modification.  */
+	      ptr = (bfd_byte *) contents + rel->r_offset + 5;
+	      code = bfd_get_8 (input_bfd, ptr);
+	      bfd_put_8 (input_bfd, (code & 0xC0) | 0x35, ptr);
+	      bfd_put_8 (input_bfd, 0, ptr + 2);
+	      /* Adjust to normal instead of PC-relative relocations. This counts
+	         on the two sets of three howtos being sequential.  */
+	      r_type = ELF32_R_TYPE (rel->r_info -= 3);
+	      howto -= 3;
+	    }
+	  /* Fall through.  */
+	case R_68K_8:
+	case R_68K_16:
+	case R_68K_32:
 	  if (bfd_link_pic (info)
 	      && r_symndx != STN_UNDEF
 	      && (input_section->flags & SEC_ALLOC) != 0
